Greeter on Wayland

plfiorini opened the pull request:
Depends on #1371 and at the moment that PR is contained here too.
Once #1371 is merged, I will rebased this on `develop`.

plfiorini committed:
Wayland display server

Introduce a new display server option to run the greeter with.
In the future this will be the default display server.

This patch contains the most important parts to get the greeter running
on Wayland, but it's not possible to select a keyboard layout.

At the moment the Wayland socket is assumed to be "wayland-0" but in a
multi-seat configuration we would have more than one greeter running,
each with a diffent socket and the assumption won't work anymore.

In the future we will figure out how to do, after all at the moment this
is considered enough for most people.

Closes: #440

plfiorini committed:
X11 display server without root privileges

Introduce an alternative display server option to run the greeter with
the X11 server as an unprivileged user.

Root privileges are required by default, but this will change in the
future.

Greeter output is forwarded to the helper process, instead of saving it
into a separate file like user sessions do.
This means the greeter output is available in the journal with the
daemon and helper logs.

Display start and stop commands are executed as sddm user and this might
break the workflow of our users.
That is the reason why we still run Xorg as root for the greeter by
default.

X11 user session always spawn a new display server without root
privileges.

Closes: #246

plfiorini committed:
Allocate VT for the display

Replace the old crude algorithm to find the next available VT with a
more reliable method using the VT_OPENQRY ioctl.

General.MinimumVT setting is now obsolete and it's no longer used.

plfiorini committed:
Restart the greeter only on crash

Keep the greeter process alive and switch to its VT instead of
restarting it, when the user session is finished.

This makes it for a very fast switch to the greeter and easier
management.

plfiorini committed:
Stop the helper process

Make sure the helper process is stopped when shutting down.

plfiorini commented:
@aleixpol Check this out

zzag commented on src/helper-x11/xorguserhelper.cpp:
```diff
@@ -0,0 +1,274 @@
+    // Create pipe for communicating with X server
+    // 0 == read from X, 1 == write to X
+    int pipeFds[2];
+    if (::pipe(pipeFds) != 0) {
```
The file descriptor for the read endpoint will be leaked to the X server process. Maybe set the O_CLOEXEC flag after creating the pipe?

```
fcntl(pipeFds[0], F_SETFD, FD_CLOEXEC);
```

zzag commented on src/helper-wayland/waylanduserhelper.cpp:
```diff
@@ -0,0 +1,166 @@
+                                     const QProcessEnvironment &env,
+                                     QProcess **p)
+{
+    auto args = QProcess::splitCommand(cmd);
```
Note that QProcess::splitCommand() was introduced in Qt 5.15, but it seems like SDDM requires Qt >= 5.8.

plfiorini replied:
it might be the time to require a newer version

tpgxyz commented:
I've compiled sddm with this PR, updated it on my OpenMandriva Lx box, updated configs, and it works with wayland greeter :)
Using it for couple of hours, from time to time do relogin, logout, swtich user etc.

aleixpol committed:
Only set QT_IM_MODULE if we have a value to give

This allows us to have it unset on Wayland so Qt defaults to text input
and the compositor can treat it as a normal application and show its
virtual keyboard infrastructure.

plfiorini committed:
Merge pull request #6 from aleixpol/feature/wayland

Only set QT_IM_MODULE if we have a value to give

Mazino-Urek commented:
Will wayland support land before Plasma 5.22?

Conan-Kudo commented:
@plfiorini Can you rebase this PR now that #1371 has been merged?
